design-documents/graph-ql/nullability.md

```diff
@@ -0,0 +1,206 @@
+   "data": {
+      "product": null
+   },
+   "errors": [
+      // details about failure in related_products field populated here
+	]
+}
+```
+
+## Backwards Compatibility
+
+- It is a safe, non-breaking change to move a field from nullable to non-nullable
+- It is an unsafe, breaking change to move a field from non-nullable to nullable
+
+With this in mind, when in doubt, the safer route is to make a field nullable, with the agreement that we'll iterate and improve schemas as we get feedback on usage from clients.
+
+**Note**: The [GraphQL reference implementation](https://github.com/graphql/graphql-js) has a `findBreakingChanges` feature that can be inspected to see a [list of all known schema changes that cause breaks for clients](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L41-L67). It [explicitly notes](https://github.com/graphql/graphql-js/blob/2232ebdef828566f3add3ed2a31709d3c1710c0e/src/utilities/findBreakingChanges.js#L461) that moving from a nullable value to a non-nullable value of the same type is safe.
+
+## Recommendations
+
+### Top-level Query fields should always be nullable
+
+Because of error propagation, if an error occurs in a non-null field on the `Query` root, the client will lose data from all other top-level queries.
+
+```graphql
+type Query {
+    # If a client queries for both fields, and one of them fails,
+    # the client will receive no data
+    product(id: ID): Product!
+    category(id: ID): Category!
+}
+```
+
+### Primary keys (id field) should always be non-nullable
+
+It very rarely makes sense to have a resource that _can_ have an ID but might not.
+
+IDs are extremely important for caching in most GraphQL clients, so it's worthwhile to be safe here.
+
+```graphql
+type Product {
+    id: ID! # Rarely makes sense for this to be nullable
+}
+```
+
+### Consider the Parent Type
+
+If you're not dealing with an `id` field or a top-level `Query` field, the most important question to ask is:
+
+**Is the parent object still usable if this field has an error?**
+
+#### Example: Parent still usable with field error
+
+```graphql
+type Product {
+    # Recommended products are not critical data on a product page, and a UI can represent
+    # a product safely without related products, so we keep the field nullable
+	recommended_products: ProductRecommendations
+}
+```
+
+#### Example: Parent not usable with field error
+
+```graphql
+type Product {
+    # A user would not be able to add a product to the cart from the Product
+    # details page if this field fails, because it may have required options.
+    # We make the field's type non-nullable
+    options: ProductOptions!
+}
+```
+
+### List Types
+
+List fields have [2 distinct forms of nullability](http://spec.graphql.org/draft/#sec-Combining-List-and-Non-Null):
+
+1. Field nullability (same as all other fields)
+2. List nullability (whether a list can have `null` items inside it)
```

Comment:
How about the following? If modified should be fixed in other places where "List nullability" is mentioned.

```suggestion
2. List item nullability (whether a list can have `null` items inside it)
```

Reply:
Good suggestion! Pushed updates